File: hangman/game.py
"""
Class for hangman games the the generator generates. Tracks
current word state, current letters guessed, target word, and
remaining guesses.
"""


class HangmanGame:
    def __init__(self, word, guesses):
        self.word = word
        self.guesses = guesses
        self.state = self.init_game_state()

    def init_game_state(self):
        return {'state': ['_'] * len(self.word), 'wrong': 0, 'guessed': set()}

    def reset_game_state(self):
        self.state = self.init_game_state()

    def process_letter_guess(self, letter):
        self.state['guessed'].add(letter)
        occurrences = [i for i, char in enumerate(self.word) if char == letter]
        for i in occurrences:
            self.state['state'][i] = letter
        self.state['wrong'] += 1 if len(occurrences) == 0 else 0

    def is_win(self):
        return ''.join(self.state['state']) == self.word and self.state['wrong'] < self.guesses

    def is_loss(self):
        return self.state['wrong'] >= self.guesses

    def game_over(self):
        return self.is_win() or self.is_loss()

    def check_for_win_status(self):
        if self.is_win():
            return ('You solved "{}" with {} guesses remaining.'.format(
                self.word,
                self.guesses - self.state['wrong']
            ))
        if self.is_loss():
            return ('You failed "{}" with {}.'.format(
                self.word,
                self.state['state']
            ))

    def print_game_state(self):
        print('--------------------------')
        print('Current state: {}'.format(self.state['state']))
        print('Guesses: {}'.format(', '.join(self.state['guessed'])))
        print('Guesses left: {}'.format(self.guesses - self.state['wrong']))
        print('--------------------------')

    def get_state(self):
        return {
            'state': ''.join(self.state['state']),
            'remaining': self.guesses - self.state['wrong'],
            'guessed': ', '.join(self.state['guessed']),
            'win': self.is_win()
        }

File: static/game.js
class Game {
	constructor(word, max_guesses) {
		this.word = word;
		this.max_guesses = max_guesses;
		this.state = this.init_game_state();
	}

	init_game_state() {
		return {
			word_state: Array(this.word.length).fill('_'),
			remaining: this.max_guesses,
			guessed: new Set(),
			status: null
		}
	}

	guess(letter) {
		this.state.guessed.add(letter);
		this.state.word_state = this.word.split('')
			.map((x, i) => x == letter || this.state.word_state[i] != '_' ? x : '_');
		this.state.remaining -= this.word.includes(letter) ? 0 : 1;
	}

	is_win() {
		return this.word == this.state.word_state.join('') && this.state.remaining > 0;
	}

	is_loss() {
		return this.state.remaining <= 0;
	}

	game_over() {
		if (this.is_win()) { this.state.status = 'win' }
		if (this.is_loss()) { this.state.status = 'loss' }
		return this.is_win() || this.is_loss();
	}

	run_with_solver(s, word_list) {
		while (!this.game_over()) {
			this.guess(s.make_guess(this.state.guessed, this.state));
		}
		return this.results;
	}

	get results() {
		if (this.state.status == 'win') {
			return 'Your bot solved "' + this.word + '" with ' 
				+ this.state.remaining + ' guesses remaining.';
		} else {
			return 'Your bot failed "' + this.word + '" with "' 
				+ this.state.word_state.join('') + '".';
		}
	}
}